# List of contents
### 01. Importing libraries and data
### 02. Creating the price range and busiest day columns
### 03. Exercise Task

# 01. Importing libraries and data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import data
path = r'C:\Users\Lawrence\Documents\Career Foundry Projects\Data Immersion\Achievement #4\2023 Instacart Basket Analysis'
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 02. Creating the price range and busiest day columns

In [3]:
# Create a subset
df = ords_prods_merged[:1000000]
df.shape

(1000000, 14)

In [4]:
# Define the function using if statements 
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [5]:
# Apply the function
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\Lawrence\AppData\Local\Temp\ipykernel_13440\2004007646.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [6]:
df['price_range'].value_counts()

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [7]:
df['prices'].max()

14.8

In [8]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Lawrence\AppData\Local\Temp\ipykernel_13440\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
C:\Users\Lawrence\AppData\Local\Temp\ipykernel_13440\1169838859.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [9]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [10]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [11]:
df['price_range_loc'].value_counts()

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [12]:
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Lawrence\AppData\Local\Temp\ipykernel_13440\953094746.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [13]:
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [14]:
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
ords_prods_merged['price_range_loc'].value_counts()

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [16]:
ords_prods_merged['orders_day_of_the_week'].value_counts(dropna = False)

orders_day_of_the_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [17]:
# Use a for-loop to classify busy days
result = []

for value in ords_prods_merged["orders_day_of_the_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [18]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [19]:
ords_prods_merged['busiest_day'] = result
ords_prods_merged['busiest_day']. value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 03. Exercise Task

In [20]:
# Step 2
# Create a new column which indicates the two busiest days and the two slowest days
result = []

for value in ords_prods_merged["orders_day_of_the_week"]:
  if value == 0:
    result.append("Busiest days")
  elif value == 1:
    result.append("Busiest days")
  elif value == 3:
    result.append("Slowest days")
  elif value == 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

ords_prods_merged['busiest_days'] = result

In [21]:
# Step 3
# Check the values of the new column for acccuracy
ords_prods_merged['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [22]:
ords_prods_merged['orders_day_of_the_week'].value_counts(dropna = False)

orders_day_of_the_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [23]:
print(6204182 + 5660230)
print(3840534 + 3783802)

11864412
7624336


The values in the new column appear to be accurate. I checked by manually adding up the corresponding numbers. 

In [24]:
# Re-examining the dataframe
ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order_number  product_id  add_to_cart_order  reordered  \
0                            NaN         196                  1          0   
1                           15.0         196                  1          1   
2                           21.0         196                  1          1   
3                           29.0         196                  1          1   
4                           28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy    Slowest days  
2  Regularly busy    Slowest days  
3      Least busy    Slowest days  
4      Least busy    Slowest days

In [25]:
# Step 4
# Create a new column to classify the busyness of periods during the day

# Examine the frequency of orders by hour of day.
ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

Based on this information, orders per hour range from 0 to 2,761,760. The following criteria will be created:

Fewest orders: 0 to 500,000 orders

Average orders: 500,001 to 2,500,000 orders

Most orders: 2,500,001+ orders


That leaves us with the following grouping:

Fewest orders: 3, 4, 2, 5, 1, 0, 6, 23

Most orders: 10, 11, 14, 15, 13, 12, 16, 9

Average orders: The remaining hours (17, 8, 18, 19, 20, 7, 21, 22)

In [26]:
# Use a for-loop to create the new column
result = []

for value in ords_prods_merged["order_hour_of_day"]:
  if value in [3, 4, 2, 5, 1, 0, 6, 23]:
    result.append("Fewest orders")
  elif value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result.append("Most orders")
  else:
    result.append("Average orders")

ords_prods_merged['busiest_period_of_day'] = result

In [27]:
# Step 5
# Print the frequency of the new column
ords_prods_merged['busiest_period_of_day'].value_counts()

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [28]:
# Export data
ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_derived.pkl'))